lib/portlayer/storage/nfs/volume.go

```diff
@@ -0,0 +1,260 @@
+	volumeStoreDir string
+
+	//handler for establishing connection to a target.
+	MH MountHandler
+
+	//nfs target for filesystem interaction
+	Target *url.URL
+
+	//Service selflink to volume store.
+	SelfLink *url.URL
+}
+
+func NewVolumeStore(op trace.Operation, storeName string, nfsTargetURL *url.URL, mount MountHandler) (*NFSv3VolumeStore, error) {
+	// XXX: Potential credential leak, maybe log less...
+	op.Infof("Creating datastore (%s) at target (%q)", storeName, nfsTargetURL.String())
+
+	target, err := mount.Mount(nfsTargetURL)
+	if err != nil {
+		return nil, err
+	}
+
+	defer mount.Unmount(target)
+
+	//we assume that nfsTargetURL.path already exists.
+	//make volumes directory
+	if _, err := target.MkDir(path.Join(nfsTargetURL.Path, VolumesDir), filePermissions); err != nil && !os.IsExist(err) {
+		return nil, err
+	}
+
+	//make metadata directory
+	if _, err := target.MkDir(path.Join(nfsTargetURL.Path, metadataDir), filePermissions); err != nil && !os.IsExist(err) {
+		return nil, err
+	}
+
+	selfLink, err := util.VolumeStoreNameToURL(storeName)
+	if err != nil {
+		return nil, err
+	}
+
+	v := &NFSv3VolumeStore{
+		Name:           storeName,
+		volumeStoreDir: nfsTargetURL.Path,
+		MH:             mount,
+		Target:         nfsTargetURL,
+		SelfLink:       selfLink,
+	}
+
+	return v, nil
+}
+
+// Returns the path to the vol relative to the given store.  The dir structure
+// for a vol in a nfs store is `<configured nfs server path>/volumes/<vol ID>/<volume contents>`.
+func (v *NFSv3VolumeStore) volDirPath(ID string) string {
+	return path.Join(v.volumeStoreDir, VolumesDir, ID)
+}
+
+func (v *NFSv3VolumeStore) volumesDir() string {
+	return path.Join(v.volumeStoreDir, VolumesDir)
+}
+
+// Returns the path to the metadata directory for a volume
+func (v *NFSv3VolumeStore) volMetadataDirPath(ID string) string {
+	return path.Join(v.volumeStoreDir, metadataDir, ID)
+}
+
+// Creates a volume directory and volume object for NFS based volumes
+func (v *NFSv3VolumeStore) VolumeCreate(op trace.Operation, ID string, store *url.URL, capacityKB uint64, info map[string][]byte) (*storage.Volume, error) {
+	target, err := v.MH.Mount(v.Target)
+	if err != nil {
+		return nil, err
+	}
+
+	defer v.MH.Unmount(target)
+
+	if _, err := target.MkDir(v.volDirPath(ID), filePermissions); err != nil {
+		return nil, err
+	}
+
+	backing := NewNFSVolumeDevice(v.Target, v.volDirPath(ID))
+
+	vol, err := storage.NewVolume(v.SelfLink, ID, info, backing)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := writeMetadata(op, v.volMetadataDirPath(ID), info, target); err != nil {
+		return nil, err
+	}
+
+	op.Infof("nfs volume (%s) successfully created on volume store (%s)", ID, v.Name)
+	return vol, nil
+}
+
+// Removes a volume and all of it's contents from the nfs store. We already know via the cache if it is in use.
+func (v *NFSv3VolumeStore) VolumeDestroy(op trace.Operation, vol *storage.Volume) error {
+	target, err := v.MH.Mount(v.Target)
```

Comment:
@hmahmood lingering mounts for our model is wasteful.  Each mount allocates a page (or set of pages) on the server.  And for our usage model, that's really unnecessary.  And depending on the server implementation, mounting can be trivial.